compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/JvmBuiltinOptimizationLowering.kt

```diff
@@ -102,6 +102,21 @@ class JvmBuiltinOptimizationLowering(val context: JvmBackendContext) : FileLower
                 expression.branches.removeIf() {
                     it.condition.isFalseConst() && isCompilerGenerated
                 }
+                if (expression.type.isBoolean() && expression.branches.size == 2) {
+                    // Replace conjunction condition with intrinsic "and" function call
+                    // the pattern for conjunction is "a && b == if (a) b else false"
+                    if (expression.branches[1].condition.isTrueConst() && expression.branches[1].result.isFalseConst()) {
```

Comment:
@neetopia Sorry, yesterday I also missed next case when debug info for 'false' would be lost and additional redundant (IFEQ) instuction would be generated:
```
fun cond() = false

fun foo() {
    if (cond())
        cond()
    else
         false
}
```
It should be checked that expression origin is IrStatementOrigin.ANDAND
Please also add custom linenumber test for that: compiler/testData/lineNumber/custom

Reply:
Good catch, the previous condition is too general, addressed.

Reply:
@neetopia Please also add assertion for expected structure (otherwise it could be conflict in future with other optimizations)